{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "12a40708",
   "metadata": {},
   "source": [
    "# Dual RNA-Sequencing Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fc3de39",
   "metadata": {},
   "source": [
    "- 24 rumen samples from an unique cohort of cows\n",
    "- Prokaryotic and Eukaryotic ribodepletion for Dual RNA-Seq\n",
    "- Sequenced at 150M paired-end reads per sample using Illumina paired-end short-reads\n",
    "- Quality checked using FastQC and MultiQC\n",
    "- Low quality reads and adaptors removed using Trimmomatic\n",
    "- Host and nonhost reads separated using kraken2 (--classified-out --unclassified-out) with a db with the bovine reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c94d497e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# commands to build host database were performed prior to the analysis below\n",
    "cd ~/Rumen_Microbiome_Genomics/1_Sequences_Guanhui/kraken2-2.1.2/\n",
    "\n",
    "# Add the PhiX genome for quality control\n",
    "kraken2-build --download-taxonomy --db Cow_and_PhiX_db\n",
    "curl -L https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/819/615/GCF_000819615.1_ViralProj14015/GCF_000819615.1_ViralProj14015_genomic.fna.gz | gunzip > phiX.fa\n",
    "kraken2-build --add-to-library phiX.fa --db Cow_and_PhiX_db --no-masking\n",
    "\n",
    "# Add cow to the library\n",
    "kraken2-build --add-to-library bosTaurus.fa --db Cow_and_PhiX_db\n",
    "\n",
    "# At the end build all downloaded libraries\n",
    "kraken2-build --build --threads 8 --db Cow_and_PhiX_db\n",
    "kraken2-build --clean --db Cow_and_PhiX_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa65c616",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import sys\n",
    "\n",
    "inpath = \"/home/hfm/Rumen_Microbiome_Genomics/1_Sequences_Guanhui/SAMPLES/\"\n",
    "outpath = \"/home/hfm/Rumen_Microbiome_Genomics/1_Sequences_Guanhui/\"\n",
    "\n",
    "\n",
    "SAMPLES=[] \n",
    "for name in glob.glob(inpath+\"*_R1_001.fixed.gz\"):\n",
    "    ID = os.path.basename(name).split(\"_R1_001\")[0]\n",
    "    if ID not in SAMPLES:\n",
    "        SAMPLES.append(ID)\n",
    "\n",
    "print('samples are:', SAMPLES, file=sys.stderr)\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        fastqc1 = expand(outpath+\"FASTQC/{sample}_R1_001.fixed_fastqc.html\", sample=SAMPLES),\n",
    "        fastqc2 = expand(outpath+\"FASTQC/{sample}_R2_001.fixed_fastqc.html\", sample=SAMPLES),\n",
    "        trimmed1 = expand(outpath+\"TRIMMED/{sample}_R1_001_trim.fixed.gz\", sample=SAMPLES),\n",
    "        trimmed2 = expand(outpath+\"TRIMMED/{sample}_R2_001_trim.fixed.gz\", sample=SAMPLES),\n",
    "        se1 = expand(outpath+\"TRIMMED/{sample}_R1_001_se.fixed.gz\", sample=SAMPLES),\n",
    "        se2 = expand(outpath+\"TRIMMED/{sample}_R2_001_se.fixed.gz\", sample=SAMPLES),\n",
    "        host1 = expand(outpath+\"host_reads/{sample}_host_reads_1.fixed.gz\", sample=SAMPLES),\n",
    "        host2 = expand(outpath+\"host_reads/{sample}_host_reads_2.fixed.gz\", sample=SAMPLES),\n",
    "        nonhost1 = expand(outpath+\"nonhost_reads/{sample}_non_host_reads_1.fixed.gz\", sample=SAMPLES),\n",
    "        nonhost2 = expand(outpath+\"nonhost_reads/{sample}_non_host_reads_2.fixed.gz\", sample=SAMPLES),\n",
    "\n",
    "\n",
    "rule make_fastqc:\n",
    "    input:\n",
    "        read1 = inpath+\"{sample}_R1_001.fixed.gz\",\n",
    "        read2 = inpath+\"{sample}_R2_001.fixed.gz\"\n",
    "    output:\n",
    "        fastqc1 = outpath+\"FASTQC/{sample}_R1_001.fixed_fastqc.html\",\n",
    "        fastqc2 = outpath+\"FASTQC/{sample}_R2_001.fixed_fastqc.html\"\n",
    "    shell:\n",
    "        \"fastqc {input}\"\n",
    "\n",
    "rule trimmomatic:\n",
    "    input:\n",
    "        read1 = inpath+\"{sample}_R1_001.fixed.gz\",\n",
    "        read2 = inpath+\"{sample}_R2_001.fixed.gz\"\n",
    "    output:\n",
    "        trimmed1 =  outpath+\"TRIMMED/{sample}_R1_001_trim.fixed.gz\",\n",
    "        trimmed2 =  outpath+\"TRIMMED/{sample}_R2_001_trim.fixed.gz\",\n",
    "        se1 = outpath+\"TRIMMED/{sample}_R1_001_se.fixed.gz\",\n",
    "        se2 = outpath+\"TRIMMED/{sample}_R2_001_se.fixed.gz\"\n",
    "    params:\n",
    "        adaptor = \"/home/hfm/Rumen_Microbiome_Genomics/1_Sequences_Guanhui/all_illumina_trimmomatic_2022Aug01.fa\"\n",
    "    conda:\n",
    "        \"trimmomatic.yaml\"\n",
    "    shell:\"\"\"\n",
    "        trimmomatic PE -quiet {input.read1} {input.read2} {output.trimmed1} {output.se1} {output.trimmed2} {output.se2}\\\n",
    "        ILLUMINACLIP:{params.adaptor}:2:40:15 LEADING:2 TRAILING:2 SLIDINGWINDOW:4:15 MINLEN:50\n",
    "        \"\"\"\n",
    "\n",
    "rule kraken_sort:\n",
    "    input:\n",
    "         trimmed1 = outpath+\"TRIMMED/{sample}_R1_001_trim.fixed.gz\",\n",
    "         trimmed2 = outpath+\"TRIMMED/{sample}_R2_001_trim.fixed.gz\"\n",
    "    output:\n",
    "         host1 = outpath+\"host_reads/{sample}_host_reads_1.fixed.gz\",\n",
    "         host2 = outpath+\"host_reads/{sample}_host_reads_2.fixed.gz\",\n",
    "         nonhost1 = outpath+\"nonhost_reads/{sample}_non_host_reads_1.fixed.gz\",\n",
    "         nonhost2 = outpath+\"nonhost_reads/{sample}_non_host_reads_2.fixed.gz\"\n",
    "    params:\n",
    "         db = \"/home/hfm/Rumen_Microbiome_Genomics/1_Sequences_Guanhui/kraken2-2.1.2/Cow_and_PhiX_db\",\n",
    "         host = outpath+\"host_reads/{sample}_host_reads#.fixed\",\n",
    "         nonhost = outpath+\"nonhost_reads/{sample}_non_host_reads#.fixed\",\n",
    "         host1 = outpath+\"host_reads/{sample}_host_reads_1.fixed\",\n",
    "         host2 = outpath+\"host_reads/{sample}_host_reads_2.fixed\",\n",
    "         nonhost1 = outpath+\"nonhost_reads/{sample}_non_host_reads_1.fixed\",\n",
    "         nonhost2 = outpath+\"nonhost_reads/{sample}_non_host_reads_2.fixed\"\n",
    "    shell:\"\"\"\n",
    "         kraken2 --threads 16 --db {params.db} --gzip-compressed --classified-out {params.host} --unclassified-out {params.nonhost} --paired {input.trimmed1} {input.trimmed2}\n",
    "         pigz -f {params.host1} > {output.host1}\n",
    "         pigz -f {params.host2} > {output.host2}\n",
    "         pigz -f {params.nonhost1} > {output.nonhost1}\n",
    "         pigz -f {params.nonhost2} > {output.nonhost2}\n",
    "         \"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tensorflow-metal-test",
   "language": "python",
   "name": "tensorflow-metal-test"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
