{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "12a40708",
   "metadata": {},
   "source": [
    "# Containment Analysis on RNA reads"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fc3de39",
   "metadata": {},
   "source": [
    "- 24 rumen samples from an unique cohort of cows\n",
    "- Prokaryotic and Eukaryotic ribodepletion for Dual RNA-Seq\n",
    "- Sequenced at 150M paired-end reads per sample using Illumina paired-end short-reads\n",
    "- QC and host reads removal performed in upstream_qc file\n",
    "- Taxonomy assignment performed using sourmash (--threshold-bp 3000 --scaled=1000 kmer=21) with the genbank db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa65c616",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import sys\n",
    "\n",
    "inpath = \"/home/hfm/Rumen_Microbiome_Genomics/1_Sequences_Guanhui/SAMPLES/\"\n",
    "outpath = \"/home/hfm/Rumen_Microbiome_Genomics/1_Sequences_Guanhui/\"\n",
    "\n",
    "\n",
    "SAMPLES=[] \n",
    "for name in glob.glob(inpath+\"*_R1_001.fixed.gz\"):\n",
    "    ID = os.path.basename(name).split(\"_R1_001\")[0]\n",
    "    if ID not in SAMPLES:\n",
    "        SAMPLES.append(ID)\n",
    "\n",
    "print('samples are:', SAMPLES, file=sys.stderr)\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        sign = expand(outpath+\"signatures/{sample}-reads.sig\", sample=SAMPLES),\n",
    "        matrix = outpath+\"compare/Noc_comp\",\n",
    "        plot = outpath+\"compare/Noc_comp.matrix.pdf\",\n",
    "        fastgather = expand(outpath+\"fastgather/all/{sample}_fastgather.csv\", sample=SAMPLES),\n",
    "        gather = expand(outpath+\"gather/all/{sample}_gather.csv\", sample=SAMPLES),\n",
    "        tax = expand(outpath+\"tax_annotated/all/{sample}_gather.with-lineages.csv\", sample=SAMPLES)\n",
    "\n",
    "\n",
    "rule sketch_signatures:\n",
    "    input:\n",
    "         read1 = outpath+\"nonhost_reads/{sample}_non_host_reads_1.fixed.gz\",\n",
    "         read2 = outpath+\"nonhost_reads/{sample}_non_host_reads_2.fixed.gz\"\n",
    "    output:\n",
    "          sign = outpath+\"signatures/{sample}-reads.sig\"\n",
    "    params:\n",
    "          merge = \"{sample}\"\n",
    "    conda:\n",
    "         \"sourmash.yaml\"\n",
    "    shell:\"\"\"\n",
    "         sourmash sketch dna -p k=21,k=31,k=51,scaled=1000,abund --merge {params.merge} -o {output.sign} {input.read1} {input.read2}\n",
    "         \"\"\"\n",
    "\n",
    "rule sourmash_compare:\n",
    "    input:\n",
    "         sign = expand(outpath+\"signatures/{sample}-reads.sig\", sample = SAMPLES)\n",
    "    output:\n",
    "         matrix = outpath+\"compare/Noc_comp\"\n",
    "    params:\n",
    "         csv = outpath+\"compare/Noc_comp.csv\"\n",
    "    conda:\n",
    "         \"sourmash.yaml\"\n",
    "    shell:\"\"\"\n",
    "         sourmash compare {input.sign} --output {output.matrix} -k 21 --csv {params.csv}\n",
    "         \"\"\"\n",
    "\n",
    "rule sourmash_plot:\n",
    "    input:\n",
    "         matrix = outpath+\"compare/Noc_comp\"\n",
    "    output:\n",
    "         plot = outpath+\"compare/Noc_comp.matrix.pdf\"\n",
    "    conda:\n",
    "         \"sourmash.yaml\"\n",
    "    shell:\"\"\"\n",
    "         sourmash plot --labels --pdf {input.matrix}\n",
    "         \"\"\"\n",
    "\n",
    "rule sourmash_fastgather:\n",
    "    input:\n",
    "         wort = outpath+\"signatures/{sample}-reads.sig\",\n",
    "         against = outpath+\"Sourmash_database/list-genbank-2022.03.txt\",\n",
    "    output:\n",
    "         csv = outpath+\"fastgather/{sample}_fastgather.csv\"\n",
    "    log:\n",
    "         outpath+\"logs/{sample}_fastgather.log\"\n",
    "    resources:\n",
    "         rayon_exclude = 1\n",
    "    threads: 128\n",
    "    conda:\n",
    "         \"sourmash.yaml\"\n",
    "    shell: \"\"\"\n",
    "          /usr/bin/time -v sourmash scripts fastgather {input.wort} {input.against} \\\n",
    "             -o {output.csv} --threshold-bp 3000 -k 21 --scaled 1000 -c {threads} 2>&1 > {log}\n",
    "           \"\"\"\n",
    "\n",
    "rule sourmash_gather:\n",
    "    input:\n",
    "         sign = outpath+\"signatures/{sample}-reads.sig\",\n",
    "         db = \"/home/hfm/Rumen_Microbiome_Genomics/1_Sequences_Guanhui/Sourmash_database/genbank-2022.03.k21.zip\",\n",
    "         picklist = outpath+\"fastgather/{sample}_fastgather.csv\"\n",
    "    output:\n",
    "         csv = outpath+\"gather/{sample}_gather.csv\"\n",
    "    conda:\n",
    "         \"sourmash.yaml\"\n",
    "    shell:\"\"\"\n",
    "         sourmash gather {input.sign} {input.db} -k 21 --no-fail-on-empty-database --threshold-bp 3000 --scaled=1000 \\\n",
    "            --picklist {input.picklist}:match_md5:md5 -o {output.csv}\n",
    "         \"\"\"\n",
    "\n",
    "rule sourmash_tax:\n",
    "    input:\n",
    "         gather = outpath+\"gather/all/{sample}_gather.csv\"\n",
    "    output:\n",
    "         tax = outpath+\"tax_annotated/all/{sample}_gather.with-lineages.csv\"\n",
    "    params:\n",
    "         db = expand(\"/home/hfm/Rumen_Microbiome_Genomics/1_Sequences_Guanhui/Sourmash_database/genbank-2022.03/genbank-2022.03-{db}.lineages.csv.gz\", db = [\"archaea\", \"bacteria\", \"fungi\", \"viral\", \"protozoal\"]),\n",
    "         dir = directory(outpath+\"tax_annotated/all/\")\n",
    "    conda:\n",
    "         \"sourmash.yaml\"\n",
    "    shell:\"\"\"\n",
    "         sourmash tax annotate -g {input.gather} -t {params.db} -o {params.dir}\n",
    "         \"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tensorflow-metal-test",
   "language": "python",
   "name": "tensorflow-metal-test"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
