{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bb374b48",
   "metadata": {},
   "source": [
    "# Sourmash Downstream"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8117a530",
   "metadata": {},
   "source": [
    "- Containement analysis performed with nonhost reads in sourmash with the genbank db\n",
    "- Taxonomy output from sourmash was used for downstream analysis performed in R using the SourmashConsumr package (https://research.arcadiascience.com/pub/resource-sourmashconsumr/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29c07dd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# install.packages(\"remotes\")\n",
    "# remotes::install_github(\"Arcadia-Science/sourmashconsumr\")\n",
    "\n",
    "library(sourmashconsumr)\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(RColorBrewer)\n",
    "library(gpuR)\n",
    "\n",
    "setwd(\"./8_Sourmash/\")\n",
    "path <- \"./8_Sourmash/\"\n",
    "list.files()\n",
    "\n",
    "metadata <- read.csv(\"Metadata.csv\")\n",
    "\n",
    "# read in the compare file\n",
    "cpath <- \"./compare/Noc_comp.csv\"\n",
    "rumen_compare_df <- read_compare_csv(cpath, sample_to_rownames = F)\n",
    "\n",
    "# Signatures\n",
    "# Listing all files in the directory\n",
    "fpaths <- list.files(path = \"./signatures/\", full.names = TRUE, pattern = \"*.sig\")\n",
    "fpaths <- sort(fpaths)\n",
    "\n",
    "# Read in the signature files\n",
    "rumen_signatures_df <- read_signature(fpaths)\n",
    "\n",
    "# Gathers\n",
    "# List all gather files in the directory\n",
    "gpaths <- list.files(path = \"./gather/gather_1000_k21/\", full.names = TRUE, pattern = \"*.csv\")\n",
    "gpaths <- sort(gpaths)\n",
    "\n",
    "# Read in the gather files\n",
    "rumen_gather_df <- read_gather(gpaths, intersect_bp_threshold = 3000)\n",
    "\n",
    "# Taxonomy\n",
    "# List all tax_annotated files in the directory\n",
    "tpaths <- list.files(path = \"./tax_annotated/tax_1000_k21/\", full.names = TRUE, pattern = \"*.csv\")\n",
    "tpaths <- sort(tpaths)\n",
    "\n",
    "# Read in the taxonomy files\n",
    "rumen_taxonomy_annotate_df <- read_taxonomy_annotate(tpaths, intersect_bp_threshold = 3000, separate_lineage = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c2be77a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Upset kmer plot to compare signatures\n",
    "rumen_signatures_df <- rumen_signatures_df %>%\n",
    "  dplyr::filter(ksize == 21)\n",
    "\n",
    "rumen_signatures_upset_df <- from_signatures_to_upset_df(signatures_df = rumen_signatures_df)\n",
    "head(rumen_signatures_upset_df)\n",
    "\n",
    "signatures_upset <- plot_signatures_upset(upset_df = rumen_signatures_upset_df,\n",
    "                                          # arguments to ComplexUpset::upset():\n",
    "                                          stripes = c(\"cyan\"),\n",
    "                                          name = \"RFI\",\n",
    "                                          #max_size = 30,\n",
    "                                          min_size = 10000000,\n",
    "                                          themes = ComplexUpset::upset_default_themes(text=element_text(size = 16, color = \"orange\"),\n",
    "                                                                                      axis.text.y = element_text(size = 13, color = \"blue\"),\n",
    "                                                                                      panel.grid = element_blank()))\n",
    "signatures_upset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ce845bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Working on Rarefaction Plots\n",
    "# Rarefaction curve\n",
    "# filter to a single k-mer size\n",
    "rumen_signatures_df <- rumen_signatures_df %>%\n",
    "  dplyr::filter(ksize == 21)\n",
    "\n",
    "rarefaction_df <- from_signatures_to_rarefaction_df(signatures_df = rumen_signatures_df)\n",
    "head(rarefaction_df)\n",
    "\n",
    "rarefaction_plt <- plot_signatures_rarefaction(rarefaction_df = rarefaction_df, fraction_of_points_to_plot = 1)\n",
    "rarefaction_plt\n",
    "\n",
    "# adds color to sample name\n",
    "rarefaction_plt +\n",
    "  ggplot2::theme_minimal() +\n",
    "  ggplot2::geom_point(ggplot2::aes(color = name))\n",
    "\n",
    "# join the metadata data frame with the rarefaction_df\n",
    "rarefaction_df <- rarefaction_df %>%\n",
    "  dplyr::left_join(metadata, by = c(\"name\" = \"Sample_ID\"))\n",
    "\n",
    "# create and modify a plot\n",
    "plot_signatures_rarefaction(rarefaction_df = rarefaction_df, fraction_of_points_to_plot = 1) +\n",
    "  ggplot2::theme_minimal() +\n",
    "  ggplot2::geom_point(ggplot2::aes(color = RFI))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3317e576",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Working on Dimension Reduction to assess sample similarities\n",
    "rumen_compare_mds_df <- make_compare_mds(rumen_compare_df)\n",
    "head(rumen_compare_mds_df)\n",
    "\n",
    "compare_plt <- plot_compare_mds(rumen_compare_mds_df)\n",
    "compare_plt +\n",
    "  ggplot2::theme_minimal()\n",
    "\n",
    "# Add metadata\n",
    "rumen_compare_mds_df %>%\n",
    "  dplyr::left_join(metadata, by = c(\"sample\" = \"Sample_ID\"))\n",
    "\n",
    "compare_mds <- plot_compare_mds(rumen_compare_mds_df) +\n",
    "  ggplot2::geom_point(ggplot2::aes(color = metadata$RFI))\n",
    "compare_mds\n",
    "ggsave(\"compare_mds_RNA.tiff\", compare_mds, dpi = 300)\n",
    "\n",
    "plot_compare_heatmap(rumen_compare_df, cexRow = 0.75, cexCol = 0.75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "251ced02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Working on gather output\n",
    "gather_classified <- plot_gather_classified(gather_df = rumen_gather_df)\n",
    "gather_classified\n",
    "ggsave(\"./z_bacteria_figures/gather_classified_RNA.tiff\", gather_classified, dpi = 300)\n",
    "\n",
    "# Upset gather plot\n",
    "rumen_gather_upset_df <- from_gather_to_upset_df(gather_df = rumen_gather_df)\n",
    "head(rumen_gather_upset_df)\n",
    "\n",
    "gather_upset <- plot_gather_upset(upset_df = rumen_gather_upset_df, \n",
    "                                  color_by_database = T, \n",
    "                                  gather_df = rumen_gather_df,\n",
    "                                  min_size = 50)\n",
    "gather_upset\n",
    "ggsave(\"gather_upset_RNA.tiff\", gather_upset, dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fac08b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Working on taxonomy output\n",
    "glom_df <-  tax_glom_taxonomy_annotate(taxonomy_annotate_df = rumen_taxonomy_annotate_df,\n",
    "                                       tax_glom_level = \"species\",\n",
    "                                       glom_var = \"f_unique_to_query\")\n",
    "\n",
    "head(glom_df)\n",
    "\n",
    "rumen_taxonomy_upset_inputs <- from_taxonomy_annotate_to_upset_inputs(taxonomy_annotate_df = rumen_taxonomy_annotate_df, \n",
    "                                                                         tax_glom_level = \"species\")\n",
    "\n",
    "plot_taxonomy_annotate_sankey(taxonomy_annotate_df = rumen_taxonomy_annotate_df, \n",
    "                              tax_glom_level = \"family\")\n",
    "ggsave(\"sankey_RNA.tiff\", dpi = 300)\n",
    "\n",
    "#### Subset per RFI\n",
    "merged_df <- merge(rumen_taxonomy_annotate_df, metadata, \n",
    "                   by.x = \"query_name\", by.y = \"Sample_ID\")\n",
    "least_df <- subset(merged_df, RFI == \"Least\")\n",
    "most_df <- subset(merged_df, RFI == \"Most\")\n",
    "\n",
    "# For least RFI\n",
    "sankey_least <- plot_taxonomy_annotate_sankey(taxonomy_annotate_df = least_df, \n",
    "                                                tax_glom_level = \"phylum\")\n",
    "sankey_least\n",
    "# ggsave(\"sankey_least_RNA.tiff\", sankey_least, dpi = 300)\n",
    "\n",
    "# For most RFI\n",
    "sankey_most <- plot_taxonomy_annotate_sankey(taxonomy_annotate_df = most_df, \n",
    "                                                    tax_glom_level = \"phylum\")\n",
    "sankey_most\n",
    "# ggsave(\"sankey_most_RNA.tiff\", sankey_most, dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5729b641",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Building a phyloseq object for multiple downstream analysis\n",
    "library(phyloseq)\n",
    "rumen_phyloseq <- from_taxonomy_annotate_to_phyloseq(taxonomy_annotate_df = rumen_taxonomy_annotate_df, \n",
    "                                                     metadata_df = metadata %>%\n",
    "                                                     tibble::column_to_rownames(\"Sample_ID\"))\n",
    "rumen_phyloseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c08ffa3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Phyloseq object for DAA at a genome level\n",
    "# Extract OTU table\n",
    "otu_table_original <- otu_table(mergedPhyla)\n",
    "\n",
    "# Transpose the OTU table\n",
    "otu_table_transposed <- t(otu_table_original)\n",
    "\n",
    "# Replace the original OTU table in the phyloseq object with the transposed table\n",
    "mergedPhyla_t <- merge_phyloseq(\n",
    "  phyloseq(otu_table(otu_table_transposed, taxa_are_rows = FALSE)),\n",
    "  tax_table(mergedPhyla),\n",
    "  sample_data(mergedPhyla)\n",
    ")\n",
    "\n",
    "### TAXONOMY (Creating the relative abundances)\n",
    "library(reshape2)\n",
    "library(phyloseq)\n",
    "ps_domain <- tax_glom(mergedPhyla_t, \"domain\", NArm=FALSE); write.csv(otu_table(ps_domain), \"./Taxonomy_levels/ps_domain.csv\")\n",
    "ps_phylum <- tax_glom(mergedPhyla_t, \"phylum\"); write.csv(otu_table(ps_phylum), \"./Taxonomy_levels/ps_phylum.csv\")\n",
    "ps_class <- tax_glom(mergedPhyla_t, \"class\"); write.csv(otu_table(ps_class), \"./Taxonomy_levels/ps_class.csv\")\n",
    "ps_order <- tax_glom(mergedPhyla_t, \"order\"); write.csv(otu_table(ps_order), \"./Taxonomy_levels/ps_order.csv\")\n",
    "ps_family <- tax_glom(mergedPhyla_t, \"family\"); write.csv(otu_table(ps_family), \"./Taxonomy_levels/ps_family.csv\")\n",
    "ps_genus <- tax_glom(mergedPhyla_t, \"genus\", NArm=FALSE); write.csv(otu_table(ps_genus), \"./Taxonomy_levels/ps_genus.csv\")\n",
    "ps_species <- tax_glom(mergedPhyla_t, \"species\", NArm=FALSE); write.csv(otu_table(ps_species), \"./Taxonomy_levels/ps_species.csv\")\n",
    "ps_strain <- tax_glom(mergedPhyla_t, \"strain\"); write.csv(otu_table(ps_strain), \"./Taxonomy_levels/ps_strain.csv\")\n",
    "\n",
    "ps_domain.r <- transform_sample_counts(ps_domain, function(x) x / sum(x)); write.csv(otu_table(ps_domain.r), \"./Taxonomy_levels/ps_domain.r.csv\"); write.csv(tax_table(ps_domain.r), \"./Taxonomy_levels/ps_domain.r_tax.csv\")\n",
    "ps_phylum.r <- transform_sample_counts(ps_phylum, function(x) x / sum(x)); write.csv(otu_table(ps_phylum.r), \"./Taxonomy_levels/ps_phylum.r.csv\"); write.csv(tax_table(ps_phylum.r), \"./Taxonomy_levels/ps_phylum.r_tax.csv\")\n",
    "ps_class.r <- transform_sample_counts(ps_class, function(x) x / sum(x)); write.csv(otu_table(ps_class.r), \"./Taxonomy_levels/ps_class.r.csv\"); write.csv(tax_table(ps_class.r), \"./Taxonomy_levels/ps_class.r_tax.csv\")\n",
    "ps_order.r <- transform_sample_counts(ps_order, function(x) x / sum(x)); write.csv(otu_table(ps_order.r), \"./Taxonomy_levels/ps_order.r.csv\"); write.csv(tax_table(ps_order.r), \"./Taxonomy_levels/ps_order.r_tax.csv\")\n",
    "ps_family.r <- transform_sample_counts(ps_family, function(x) x / sum(x)); write.csv(otu_table(ps_family.r), \"./Taxonomy_levels/ps_family.r.csv\"); write.csv(tax_table(ps_family.r), \"./Taxonomy_levels/ps_family.r_tax.csv\")\n",
    "ps_genus.r <- transform_sample_counts(ps_genus, function(x) x / sum(x)); write.csv(otu_table(ps_genus.r), \"./Taxonomy_levels/ps_genus.r.csv\"); write.csv(tax_table(ps_genus.r), \"./Taxonomy_levels/ps_genus.r_tax.csv\")\n",
    "ps_species.r <- transform_sample_counts(ps_species, function(x) x / sum(x)); write.csv(otu_table(ps_species.r), \"./Taxonomy_levels/ps_species.r.csv\"); write.csv(tax_table(ps_species.r), \"./Taxonomy_levels/ps_species.r_tax.csv\")\n",
    "ps_strain.r <- transform_sample_counts(ps_strain, function(x) x / sum(x)); write.csv(otu_table(ps_strain.r), \"./Taxonomy_levels/ps_strain.r.csv\"); write.csv(tax_table(ps_strain.r), \"./Taxonomy_levels/ps_strain.r_tax.csv\")\n",
    "\n",
    "# CLR transformation on phyloseq object\n",
    "mergedPhyla_t_clr<-microbiome::transform(mergedPhyla_t, \"clr\")\n",
    "domain_clr<-microbiome::transform(ps_domain, \"clr\")\n",
    "phylum_clr<-microbiome::transform(ps_phylum, \"clr\")\n",
    "class_clr<-microbiome::transform(ps_class, \"clr\")\n",
    "order_clr<-microbiome::transform(ps_order, \"clr\")\n",
    "family_clr<-microbiome::transform(ps_family, \"clr\")\n",
    "genus_clr<-microbiome::transform(ps_genus, \"clr\")\n",
    "species_clr<-microbiome::transform(ps_species, \"clr\")\n",
    "strain_clr<-microbiome::transform(ps_strain, \"clr\")\n",
    "\n",
    "write.csv(otu_table(mergedPhyla_t_clr), \"./Taxonomy_levels/ps_domain_clr.csv\")\n",
    "write.csv(otu_table(domain_clr), \"./Taxonomy_levels/ps_domain_clr.csv\")\n",
    "write.csv(otu_table(phylum_clr), \"./Taxonomy_levels/ps_phylum_clr.csv\")\n",
    "write.csv(otu_table(class_clr), \"./Taxonomy_levels/ps_class_clr.csv\")\n",
    "write.csv(otu_table(order_clr), \"./Taxonomy_levels/ps_order_clr.csv\")\n",
    "write.csv(otu_table(family_clr), \"./Taxonomy_levels/ps_family_clr.csv\")\n",
    "write.csv(otu_table(genus_clr), \"./Taxonomy_levels/ps_genus_clr.csv\")\n",
    "write.csv(otu_table(species_clr), \"./Taxonomy_levels/ps_species_clr.csv\")\n",
    "write.csv(otu_table(strain_clr), \"./Taxonomy_levels/ps_strain_clr.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9905daa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# At this point, tax_table(ps_species.r) was used to build the \n",
    "# phylogenomic tree present in the study for archaea, bacteria, and eukaryotes.\n",
    "# The viral taxonomy tree was build with assembled viral contigs (vOTUs) that contained RdRP"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tensorflow-metal-test",
   "language": "python",
   "name": "tensorflow-metal-test"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
